File: src/main/java/gargoyle/ct/sound/AudioClipImpl.java
package gargoyle.ct.sound;

import gargoyle.ct.log.Log;
import gargoyle.ct.util.Defend;
import org.jetbrains.annotations.NotNull;

import javax.sound.sampled.*;
import java.io.IOException;
import java.net.URL;

public class AudioClipImpl implements AudioClip {
    boolean playing;
    private Clip clip;
    private boolean released;

    public AudioClipImpl(URL url) {
        Defend.notNull(url, "null audio clip");
        try {
            AudioInputStream stream = AudioSystem.getAudioInputStream(url);
            clip = AudioSystem.getClip();
            clip.open(stream);
            clip.addLineListener(new AudioClipListener());
            released = true;
        } catch (@NotNull IOException | UnsupportedAudioFileException | LineUnavailableException exc) {
            Log.error(exc.getLocalizedMessage(), exc);
            released = false;
        }
    }

    public boolean isReleased() {
        return released;
    }

    @Override
    public void loop() {
        play();
    }

    @Override
    public void play() {
        play(true);
    }

    public void play(boolean breakOld) {
        if (released) {
            if (breakOld) {
                clip.stop();
                clip.setFramePosition(0);
                clip.start();
                playing = true;
            } else if (!playing) {
                clip.setFramePosition(0);
                clip.start();
                playing = true;
            }
        }
    }

    public boolean isPlaying() {
        return playing;
    }

    @Override
    public void stop() {
        if (playing) {
            clip.stop();
        }
    }

    private class AudioClipListener implements LineListener {
        @Override
        public void update(@NotNull LineEvent event) {
            if (event.getType() == LineEvent.Type.STOP) {
                playing = false;
            }
        }
    }
}

File: src/main/kotlin/gargoyle/ct/sound/AudioClipImpl.kt
package gargoyle.ct.sound

import gargoyle.ct.log.Log
import gargoyle.ct.util.Defend
import java.io.IOException
import java.net.URL
import javax.sound.sampled.AudioSystem
import javax.sound.sampled.Clip
import javax.sound.sampled.LineEvent
import javax.sound.sampled.LineListener
import javax.sound.sampled.LineUnavailableException
import javax.sound.sampled.UnsupportedAudioFileException

class AudioClipImpl(url: URL) : AudioClip {
    var isPlaying = false
    private lateinit var clip: Clip
    var isReleased = false

    init {
        Defend.notNull(url, "null audio clip")
        try {
            val stream = AudioSystem.getAudioInputStream(url)
            clip = AudioSystem.getClip()
            clip.open(stream)
            clip.addLineListener(AudioClipListener())
            isReleased = true
        } catch (exc: IOException) {
            Log.error(exc.localizedMessage, exc)
            isReleased = false
        } catch (exc: UnsupportedAudioFileException) {
            Log.error(exc.localizedMessage, exc)
            isReleased = false
        } catch (exc: LineUnavailableException) {
            Log.error(exc.localizedMessage, exc)
            isReleased = false
        }
    }

    override fun loop() {
        play()
    }

    override fun play() {
        play(true)
    }

    fun play(breakOld: Boolean) {
        if (isReleased) {
            if (breakOld) {
                clip.stop()
                clip.framePosition = 0
                clip.start()
                isPlaying = true
            } else if (!isPlaying) {
                clip.framePosition = 0
                clip.start()
                isPlaying = true
            }
        }
    }

    override fun stop() {
        if (isPlaying) {
            clip.stop()
        }
    }

    private inner class AudioClipListener : LineListener {
        override fun update(event: LineEvent) {
            if (event.type === LineEvent.Type.STOP) {
                isPlaying = false
            }
        }
    }
}
